src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -1305,198 +1305,224 @@ protected override void HandlePauseRequestAsync(IRequestResponder<PauseArguments
-                            Column = textPosition.Column,
-                            ModuleId = moduleId
-                        });
-                    }
+                            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+                            try
+                            {
+                                CancellationToken cancellationToken = cancellationTokenSource.Token;
+                                Task[] frameTasks = response.StackFrames.Select(frame => Task.Run(() =>
+                                {
+                                    cancellationToken.ThrowIfCancellationRequested();
+                                    if (frame.Source != null && !File.Exists(frame.Source.Path))
+                                    {
+                                        cancellationToken.ThrowIfCancellationRequested();
+                                        // mark the stack frame to be skipped by default
+                                        frame.Source.PresentationHint = Source.PresentationHintValue.Deemphasize;
+                                    }
+                                })).ToArray();
 
-                    if (!IsClientVS)
-                    {
-                        Parallel.ForEach(response.StackFrames, frame =>
-                        {
-                            if (frame.Source == null)
-                                return;
-                            var task = new Task<bool>(() => File.Exists(frame.Source.Path));
-                            task.Start();
-                            if (task.Wait(500) && !task.Result)
+                                Task allFramesComplete = Task.WhenAll(frameTasks);
+                                Task timeoutTask = Task.Delay(20000, cancellationToken);
+                                try
+                                {
+                                    await Task.WhenAny(allFramesComplete, timeoutTask);
+                                }
+                                catch (Exception e)
+                                {
+                                    Debug.Fail("Ignoring exception computing frame state: " + e.ToString());
+                                }
+                            }
+                            finally
                             {
-                                // mark the stack frame to be skipped by default
-                                frame.Source.PresentationHint = Source.PresentationHintValue.Deemphasize;
+                                cancellationTokenSource.Cancel();
+                                cancellationTokenSource.Dispose();
                             }
-                        });
-                    }
+                        }
 
-                    response.TotalFrames = (int)frameEnumInfo.TotalFrames;
+                        response.TotalFrames = (int)frameEnumInfo.TotalFrames;
+                    }
                 }
-            }
```

Comment:
This function is one of the largest and most deeply nested ones.
If you touch all lines anyway maybe it could be split up and/or the nesting reduced.